### ISSUE TITLE
Policy based volume provisioning failed due to VM Name maxed out permitted length

### ISSUE THREAD

**Issue by divyenpatel** [vmware/kubernetes#176]
We observed this issue yesterday.

if kubernetes cluster name (VM Folder Name including path) is set within 20 chars, we do not have any issue.

Task Error: 'vsphere-k8s-dev-kubernetes-cluster-dynamic-pvc-f056635c-57bd-11e7-962b-0050569abc19' is invalid or exceeds the maximum number of characters permitted.

<img width="974" alt="vm-name-maxed-out" src="https://user-images.githubusercontent.com/22985595/27485900-8c729038-57e3-11e7-8152-ad9eceadf6f8.png">

This should be patched with priority in 1.7 release.

Ref: https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/vsphere/vsphere.go#L1404

cc: @BaluDontu  @tusharnt @rohitjogvmw @luomiao

**Comment by divyenpatel** [comment]
@rohitjogvmw This can be fixed quickly if we use vmhash as below instead of vmname.
Before making call to the VC we can convert this vmname to vmhash and this should resolve the issue.

```
package main

import (
    "crypto/md5"
    "encoding/hex"
    "fmt"
)

func main() {
	var vmname = "vsphere-k8s-dev-kubernetes-cluster-dynamic-pvc-f056635c-57bd-11e7-962b-0050569abc19"
	hash := md5.Sum([]byte(vmname))
	vmhash := hex.EncodeToString(hash[:])
	fmt.Printf(vmhash)
}
```

**Comment by divyenpatel** [comment]
@rohitjogvmw  PR - https://github.com/kubernetes/kubernetes/pull/50722 is merged, we need to cherry pick this PR to 1.7 branch.
Due to refactoring we can not apply all the changes but in place code change in 1.7 branch should be allowed.

**Comment by divyenpatel** [comment]
#51079 Merged to release 1.7. 
Closing this issue.
